torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,809 @@
+    FullyShardedDataParallel is commonly shorten to FSDP.
+    .. _`Xu et al.`: https://arxiv.org/abs/2004.13336
+    .. _DeepSpeed: https://www.deepspeed.ai/
+    Usage::
+        import torch
+        from torch.distributed._fsdp import FullyShardedDataParallel as FSDP
+        torch.cuda.set_device(device_id)
+        sharded_module = FSDP(my_module)
+        optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        x = sharded_module(x, y=3, z=torch.Tensor([1]))
+        loss = x.sum()
+        loss.backward()
+        optim.step()
+    .. warning::
+        The optimizer must be initialized *after* the module has been wrapped,
+        since FSDP will shard parameters in-place and this will break any
+        previously initialized optimizers.
+    Args:
+        module (nn.Module):
+            module to be wrapped with FSDP.
+        process_group (Optional):
+            process group for sharding
+    """
+
+    def __init__(
+        self,
+        module: nn.Module,
+        process_group: Optional[ProcessGroup] = None,
+    ):
+        init_start = time.time()
+        super().__init__()
+        self.module = module
+        self.process_group = process_group or _get_default_group()
+        self.rank = self.process_group.rank()
+        self.world_size = self.process_group.size()
+        # device for computation, if module is on GPU, use module.device;
+        # if module is on CPU, use current device;
+        self.compute_device = _get_default_cuda_device(module)
+        self.compute_dtype = next(module.parameters()).dtype
+
+        # setting two factors to avoid underflow and overflow
+        self.gradient_predivide_factor: float = self._get_gradient_predivide_factor(self.world_size)
+        self.gradient_postdivide_factor: float = self.world_size / self.gradient_predivide_factor
+
+        self.numel_padded_per_param: List[int] = []
+
+        # Only handle params which are not already sharded. This enables
+        # sharding individual layers of a Module, with an outer wrapper to
+        # shard any leftover parameters.
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
+                params.append(param)
+        self.params = params
+
+        # Shard module parameters in place
+        self._shard_parameters_()
+
+        # Make sure all parameters are sharded.
+        for n, p in self.named_parameters():
+            assert hasattr(p, "_is_sharded"), f"found unsharded parameter: {n} ; {p.size()}"
+
+        self._reset_lazy_init()
+
+        # Enum to indicate if we're in the forward/backward pass, idle, etc.
+        self.training_state = TrainingState.IDLE
+
+        # Flag to guard against preparing gradients multiple times per backward pass.
+        self._pre_backward_hook_has_run = False
+
+        init_end = time.time()
+
+        logging.debug(
```

Comment:
Curious, why this is in FSDP code? Looks like users can simply measure the `FSDP()` call to get this number?

Reply:
we can remove it, we can revisit it when we are working on debugging tool